// @biom3/react/src/components/Banner/styles.ts
import { BiomeTheme } from '../../types/theme';
import { BannerVariant } from './types';

export const bannerSx = {
  d: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'flex-start',
  alignItems: 'center',
  width: '100%',
  minHeight: 'base.spacing.x16',
  p: 'base.spacing.x4',
  columnGap: 'base.spacing.x2',
  rowGap: 'base.spacing.x4',
  paddingLeft: 'base.spacing.x14',
  position: 'relative',
  boxShadow: 'base.shadow.300',
  brad: 'base.borderRadius.x6',
};

export const rightHandButtonsContainerSx = {
  display: 'flex',
  gap: 'base.spacing.x2',
  ml: 'auto',
};

export const getBannerColoring = (variant: BannerVariant) => {
  switch (variant) {
    case 'success':
    case 'fatal':
    case 'guidance':
    case 'attention':
    case 'destructive':
      return { bg: `base.color.status.${variant}.dim` };

    case 'standard':
    default:
      return { bg: 'base.color.neutral.500' };
  }
};

export const shouldInvertColoring = (
  variant: BannerVariant,
  { base }: BiomeTheme,
) => {
  const isDarkMode = base.colorMode === 'lightOnDark';
  const isStandardVariant = variant === 'standard';
  const invertButtonColoring = isDarkMode && !isStandardVariant;
  return invertButtonColoring;
};
