// @biom3/react/src/components/Banner/types.ts
import { Colors } from '@biom3/design-tokens';
import { DistributiveOmit } from '@emotion/react';
import { ReactElement } from 'react';

import {
  ButtConProps,
  ButtonProps,
  ButtonVariant,
} from '../../types/clickable';
import { BodyProps } from '../../types/text';
import { BoxProps } from '../Box';

export const DEFAULT_BANNER_VARIANT = 'standard';
export type BannerVariant =
  | keyof Colors['status']
  | typeof DEFAULT_BANNER_VARIANT;
export type BannerProps = BoxProps & {
  variant?: BannerVariant;
  forceMultilineText?: boolean;
};
export type BannerTextProps<RC extends ReactElement | undefined = undefined> =
  BodyProps<RC> & { variant?: BannerVariant };

type StrippedDownVariants = Exclude<
  ButtonVariant,
  | 'secondary'
  | 'secondary/destructive'
  | 'tertiary/inverse'
  | 'primary/inverse'
  | 'primary/destructive'
>;

export type BannerRightButtConProps<
  RC extends ReactElement | undefined = undefined,
> = DistributiveOmit<ButtConProps<RC>, 'variant' | 'size'> & {
  bannerVariant?: BannerVariant;
  variant?: StrippedDownVariants;
};

export type BannerRightButtonProps<
  RC extends ReactElement | undefined = undefined,
> = DistributiveOmit<ButtonProps<RC>, 'variant' | 'size'> & {
  bannerVariant?: BannerVariant;
  variant?: StrippedDownVariants;
};
