// @biom3/react/src/components/Card/styles.ts
export const baseContainerStyles = {
  flexShrink: 0,
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
  width: '100%',
  brad: 'base.borderRadius.x4',
  overflow: 'hidden',
  position: 'relative',
  boxShadow: 'base.shadow.100',
  transitionProperty: 'box-shadow, translate',
  transitionDuration: 'base.motion.normal.fast.cssDuration',
  transitionTimingFunction: 'base.motion.normal.fast.cssEase',

  '&.Card--selected': {
    border: '2px solid',
    borderColor: 'base.color.translucent.standard.1000',

    '&::after': {
      content: '""',
      pos: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      brad: 'base.borderRadius.x3',
      border: '2px solid transparent',
      pointerEvents: 'none',
      userSelect: 'none',
    },
  },

  '& .CardAssetImage': {
    overflow: 'hidden',
  },

  '& .AspectRatioImage__inner__image': {
    scale: '1.0',
    transitionProperty: 'scale',
    transitionDuration: 'base.motion.normal.fast.cssDuration',
    transitionTimingFunction: 'base.motion.normal.fast.cssEase',
  },

  '&:hover:not(article)': {
    boxShadow: 'base.shadow.300',
    translate: '0 -8px',

    '& .AspectRatioImage__inner__image': {
      scale: '1.05',
    },
  },

  '&:not(article):active': {
    boxShadow: 'base.shadow.200',
    translate: '0 -4px',
  },

  '&.Card--elevation1': {
    bgc: 'base.color.neutral.900',

    '&:hover:not(article)': {
      bgc: 'base.color.neutral.700',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.700',
      },
    },

    '&:not(article):active': {
      bgc: 'base.color.neutral.800',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.800',
      },
    },

    '&.Card--selected&::after': {
      borderColor: 'base.color.neutral.900',
    },
  },

  '&.Card--elevation2': {
    bgc: 'base.color.neutral.800',

    '&:hover:not(article)': {
      bgc: 'base.color.neutral.600',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.600',
      },
    },

    '&:not(article):active': {
      bgc: 'base.color.neutral.700',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.700',
      },
    },

    '&.Card--selected&::after': {
      borderColor: 'base.color.neutral.800',
    },
  },

  '&.Card--elevation3': {
    bgc: 'base.color.neutral.700',

    '&:hover:not(article)': {
      bgc: 'base.color.neutral.500',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.500',
      },
    },

    '&:not(article):active': {
      bgc: 'base.color.neutral.600',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.600',
      },
    },

    '&.Card--selected&::after': {
      borderColor: 'base.color.neutral.700',
    },
  },

  '&.Card--elevation4': {
    bgc: 'base.color.neutral.600',

    '&:hover:not(article)': {
      bgc: 'base.color.neutral.500',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.500',
      },
    },

    '&:not(article):active': {
      bgc: 'base.color.neutral.500',

      '&.Card--selected&::after': {
        borderColor: 'base.color.neutral.500',
      },
    },

    '&.Card--selected&::after': {
      borderColor: 'base.color.neutral.600',
    },
  },

  '&.Card--elevation5': {
    bgc: 'base.color.neutral.500',

    '&.Card--selected&::after': {
      borderColor: 'base.color.neutral.500',
    },
  },
};

export const baseFramedStyles = {
  d: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  position: 'absolute',
  top: '-32px',
  left: 'base.spacing.x6',
  bg: 'base.color.neutral.900',
  overflow: 'hidden',
  width: 'base.icon.size[500]',
  height: 'base.icon.size[500]',
  borderRadius: 'base.borderRadius.x4',
};
