// @biom3/react/src/overlays/DrawerOverlay/styles.ts
import { DesignTokens } from '@biom3/design-tokens';
import { Properties } from 'csstype';

import { DrawerPosition, DrawerSize } from '../../types/overlays';

export function getDrawerSize(size: DrawerSize, position: DrawerPosition) {
  switch (size) {
    case 'half':
      return '50%';
    case 'quarter':
      return '25%';
    case 'full':
      return /bottom/.test(position) ? 'calc(100% - 48px)' : '100%';
    case 'threeQuarter':
    default:
      return '75%';
  }
}

export const BASE_TRANSLATE = '32px';

export function getDrawerStyles(
  position: DrawerPosition,
  size: DrawerSize,
  base: DesignTokens['base'],
) {
  const styles: Properties = {};
  switch (position) {
    case 'right':
    case 'left':
      styles.height = '100%';
      styles.width = getDrawerSize(size, position);
      break;

    case 'bottom':
    default:
      styles.width = '100%';
      styles.height = getDrawerSize(size, position);
      break;
  }

  switch (position) {
    case 'left':
      styles.left = '0px';
      styles.top = '0px';
      styles.boxShadow = `-${BASE_TRANSLATE} 0 0 ${base.color.neutral[700]}`;
      styles.borderRadius = `0 ${base.borderRadius.x10} ${base.borderRadius.x10} 0`;
      break;

    case 'right':
      styles.top = '0px';
      styles.right = '0px';
      styles.boxShadow = `${BASE_TRANSLATE} 0 0 ${base.color.neutral[700]}`;
      styles.borderRadius = `${base.borderRadius.x10} 0 0 ${base.borderRadius.x10}`;
      break;

    case 'bottom':
    default:
      styles.bottom = '0px';
      styles.left = '0px';
      styles.boxShadow = `0 ${BASE_TRANSLATE} 0 ${base.color.neutral[700]}`;
      styles.borderRadius = `${base.borderRadius.x10} ${base.borderRadius.x10} 0 0`;
      break;
  }

  return styles;
}

export function getDrawerCloseIcon(position: DrawerPosition) {
  switch (position) {
    case 'left':
      return 'ChevronBackward';

    case 'right':
      return 'ChevronForward';

    case 'bottom':
    default:
      return 'ChevronExpand';
  }
}

export const RIGHT_TRANSLATE = BASE_TRANSLATE;
export const LEFT_TRANSLATE = `-${BASE_TRANSLATE}`;
export const BOTTOM_TRANSLATE = BASE_TRANSLATE;
export const NO_TRANSLATE = '0px';
