// @biom3/react/src/components/Clickable/Buttons/styles/shared.ts
import { BiomeTheme } from '../../../../types/theme';

export const smallButtonStaticStyles = ({ base }: BiomeTheme) => ({
  height: base.spacing.x8,
  fontSize: base.text.body.xSmall.regular.fontSize,
  fontWeight: base.text.body.medium.bold.fontWeight,
  fontFamily: base.font.family.body.primary,
  borderRadius: base.borderRadius.x15,
  paddingX: base.spacing.x4,
  gap: '2px',
});

export const mediumButtonStaticStyles = ({ base }: BiomeTheme) => ({
  height: base.spacing.x12,
  fontSize: base.text.body.medium.regular.fontSize,
  fontWeight: base.text.body.medium.bold.fontWeight,
  fontFamily: base.font.family.body.primary,
  borderRadius: base.borderRadius.x15,
  paddingX: base.spacing.x6,
  gap: base.spacing.x1,
});

export const largeButtonStaticStyles = ({ base }: BiomeTheme) => ({
  height: base.spacing.x16,
  fontSize: base.text.body.medium.regular.fontSize,
  fontWeight: base.text.body.medium.bold.fontWeight,
  fontFamily: base.font.family.body.primary,
  borderRadius: base.borderRadius.x20,
  paddingX: base.spacing.x8,
  gap: base.spacing.x2,
});
