// @biom3/react/.eslintrc.js
module.exports = {
  root: false,
  extends: ['custom'],
  rules: {
    '@next/next/no-img-element': 'off',
    'guard-for-in': 'off',
    'no-restricted-syntax': ['warn', 'ForInStatement'],
    'import/no-cycle': 'warn',
    'react/jsx-no-useless-fragment': 'warn',
    'react/static-property-placement': 'off',
    'jsx-a11y/anchor-has-content': 'off',
    'jsx-a11y/anchor-is-valid ': 'off',
    'jsx-a11y/control-has-associated-label': 'off',
    'react/button-has-type': 'off',
    'jsx-a11y/heading-has-content': 'off',
    'jsx-a11y/anchor-is-valid': 'off',
    'jsx-a11y/label-has-associated-control': 'off',
    'import/no-extraneous-dependencies': [
      'error',
      { devDependencies: ['**/*.test.tsx'] },
    ],
    'no-underscore-dangle': ['error', { allow: ['__html'] }],
  },
  parserOptions: {
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
  },
  globals: {
    Cypress: 'readonly',
    cy: 'readonly',
    describe: 'readonly',
    it: 'readonly',
    beforeEach: 'readonly',
    afterEach: 'readonly',
    expect: 'readonly',
    context: 'readonly',
    globalThis: 'readonly',
    JSX: 'readonly',
  },
  ignorePatterns: ['*.config.ts', 'jest-setup.ts', 'cypress/**/*.ts'],
};
