// @biom3/react/src/components/RecycleView/style.ts
import { Interpolation } from '@emotion/react';
import { CSSProperties } from 'react';

import { SxProps } from '../../types';
import { BiomeTheme } from '../../types/theme';

export const containerSx: SxProps = {
  flexGrow: 1,
  width: '100%',
  height: '100%',
  maxHeight: '100vh',
  display: 'flex',
  flexDirection: 'column',
  position: 'relative',
};

export const overflowCss = (
  padForCardItems: boolean,
): Interpolation<BiomeTheme> => ({
  flex: 1,
  overflow: 'hidden',
  overflowY: 'scroll',
  ...(padForCardItems ? { paddingLeft: '10px', paddingRight: '10px' } : {}),
});

export const gridCss = (
  padForCardItems: boolean,
  gridGap: number,
): Interpolation<BiomeTheme> => ({
  display: 'flex',
  flexWrap: 'wrap',
  gap: `${gridGap}px`,
  ...(padForCardItems ? { paddingTop: '12px', paddingBottom: '12px' } : {}),
});

export const testItemCss: Interpolation<BiomeTheme> = {
  position: 'absolute',
  top: '0',
  left: '0',
  opacity: 0,
  pointerEvents: 'none',
};

export const loadingUiSx: SxProps = {
  width: '100%',
  padding: 'base.spacing.x5',
  display: 'flex',
  justifyContent: 'center',
};

export const sentinalCss: CSSProperties = {
  pointerEvents: 'none',
  position: 'absolute',
  left: 0,
  width: '100%',
  height: '200px',
  // @TODO: un-comment this to make it easier to debug UI issues
  // background: 'red',
  // opacity: 0.6,
};
