// @biom3/react/src/components/Toggle/styles.ts
import { Properties } from 'csstype';

export const containerSx = {
  width: 'base.spacing.x12',
  height: 'base.spacing.x7',
  position: 'relative',
  d: 'flex',
  alignItems: 'center',
  justifyContent: 'flexStart',

  '&:disabled + .toggleContainer': {
    boxShadow: 'none',
  },

  'input:hover:not(:disabled) + .toggleContainer': {
    boxShadow: 'base.shadow.300',
    backgroundColor: 'base.color.translucent.emphasis.200',
  },

  'input:hover:not(:disabled) + .toggleContainer:before': {
    borderColor: 'base.color.translucent.standard.1000',
  },

  'input:hover:not(:disabled):not(checked) + .toggleContainer > .toggleHandle':
    {
      transform: 'translateX(6px)',
      boxShadow: 'base.shadow.300',
    },

  // Checked state styles:
  'input:checked + .toggleContainer': {
    background: 'base.color.status.success.bright',
    boxShadow: 'none',
  },

  'input:checked + .toggleContainer::before': {
    borderColor: 'base.color.translucent.standard.1000',
    borderWidth: 'base.border.size.200',
  },

  'input:checked:hover:not(:disabled) + .toggleContainer': {
    background: 'base.color.status.success.bright',
    boxShadow: 'base.shadow[200]',
  },

  'input:checked:not(:disabled):hover + .toggleContainer::before': {
    borderColor: 'base.color.translucent.standard.1000',
    borderWidth: 'base.border.size.100',
  },

  'input:checked + .toggleContainer > .toggleHandle': {
    transform: 'translateX(22px)',
  },

  'input:checked + .toggleContainer > .toggleHandle > .toggleHandle__tickIcon':
    {
      opacity: '1',
    },

  'input:hover:checked:not(:disabled) + .toggleContainer > .toggleHandle': {
    transform: 'translateX(20px)',
  },
};

export const toggleContainerSx = {
  position: 'relative',
  pointerEvents: 'none',
  width: 'base.spacing.x12',
  height: 'base.spacing.x7',
  borderRadius: 'base.borderRadius.x12',
  backgroundColor: 'base.color.translucent.emphasis.100',
  d: 'flex',
  alignItems: 'center',

  '&::before': {
    content: '""',
    position: 'absolute',
    top: '0',
    left: '0',
    width: '100%',
    height: '100%',
    borderRadius: 'base.borderRadius.x12',
    borderWidth: 'base.border.size.100',
    borderStyle: 'solid',
    borderColor: 'base.color.translucent.standard.500',
  },
};

export const inputCss = {
  position: 'absolute',
  top: '0',
  left: '0',
  width: '100%',
  height: '100%',
} satisfies Properties;

export const handleSx = {
  w: '22px',
  h: '22px',
  d: 'flex',
  position: 'relative',
  alignItems: 'center',
  justifyContent: 'center',
  borderRadius: '50%',
  backgroundColor: 'base.color.translucent.standard.1000',
  boxShadow: 'base.shadow.100',
  transition: '200ms ease-in-out',
  transitionProperty: 'transform, box-shadow',
  pointerEvents: 'none',
  transform: 'translateX(3px)',
};

export const tickSx = {
  fill: 'base.color.text.inverse.primary',
  opacity: '0',
};
