// @biom3/react/src/components/Radio/styles.ts
import { BiomeTheme } from '../../types/theme';

export const containerSx = {
  width: 'base.spacing.x5',
  height: 'base.spacing.x5',
  d: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  position: 'relative',
};

export const inputCssStyles = (theme: BiomeTheme) => ({
  '&:disabled + .radioContainer': {
    boxShadow: 'none',
  },

  '&:checked + .radioContainer': {
    borderWidth: theme.base.border.size[200],
    borderColor: theme.base.color.text.primary,
  },

  '&:checked + .radioContainer > .checkCircle': {
    opacity: '1',
  },

  '&:hover:not(:disabled) + .radioContainer': {
    borderWidth: theme.base.border.size[200],
  },

  '&:checked:hover:not(:disabled) + .radioContainer': {
    borderColor: theme.base.color.text.secondary,
    borderWidth: theme.base.border.size[100],
  },

  ':checked:not(:disabled):hover + .radioContainer > .checkCircle': {
    width: '6px',
    height: '6px',
  },
});

export const radioContainerSx = {
  borderStyle: 'solid',
  borderWidth: 'base.border.size[100]',
  borderColor: 'base.color.text.secondary',
  borderRadius: 'base.borderRadius.x6',
  boxShadow: 'base.shadow[100]',
  width: 'base.spacing.x5',
  height: 'base.spacing.x5',
  d: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
};

export const checkSx = {
  position: 'relative',
  pointerEvents: 'none',
  width: '10px',
  height: '10px',
  borderRadius: '50%',
  backgroundColor: 'base.color.text.primary',
  opacity: '0',
};
