// @biom3/react/src/utils/dummyImages.ts
export const DUMMY_RASTER_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/demo-images/frog.png';
export const DUMMY_VECTOR_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/currency-icons/currency--eth.svg';
export const DUMMY_RASTER_IMAGE_2_URL =
  'https://biome.immutable.com/hosted-assets/demo-images/tree-dude.jpg';
export const DUMMY_RASTER_IMAGE_3_URL =
  'https://biome.immutable.com/hosted-assets/demo-images/pastelle-gradients.png';
export const IMX_TOKEN_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/currency-icons/currency--imx.svg';
export const GODS_TOKEN_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/currency-icons/currency--gods.svg';
export const ETH_TOKEN_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/currency-icons/currency--eth.svg';
export const GOG_TOKEN_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/currency-icons/currency--gog.svg';
export const DARK_ON_LIGHT_FALLBACK_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/images/defaultTokenImage--darkOnLight.svg';
export const LIGHT_ON_DARK_FALLBACK_IMAGE_URL =
  'https://biome.immutable.com/hosted-assets/images/defaultTokenImage--lightOnDark.svg';
export const DUMMY_RIVE_ANIMATION_URL =
  'https://cdn.rive.app/animations/vehicles.riv';
