// @biom3/react/src/components/Badge/styles.ts
import { keyframes } from '@emotion/react';

import { BadgeVariant } from './shared';

const dotPulseAnimation = keyframes`
  0% {
    scale: 1;
    opacity: 0.8;  
  }
  100% {
    scale: 2.8;
    opacity: 0
  }
`;

const contentPulseAnimation = keyframes`
  0% {
    scale: 1;
    opacity: 0.8;  
  }
  100% {
    scale: 1.5 1.9;
    opacity: 0
  }
`;

export function getBadgeAnimationSx(isAnimated: boolean, hasContent: boolean) {
  return isAnimated
    ? {
        '&:before': {
          animation: `${
            hasContent ? contentPulseAnimation : dotPulseAnimation
          } infinite`,
          animationDuration: '1.3s',
          animationTimingFunction: 'base.motion.normal.gentle.cssEase',
        },
      }
    : {};
}

export const baseBadgeSx = {
  position: 'relative',
  flexShrink: '0',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  px: 'base.spacing.x2',
  py: '1px',
  brad: 'base.spacing.x6',

  '&:before': {
    content: "''",
    position: 'absolute',
    top: '0',
    left: '0',
    right: '0',
    bottom: '0',
    brad: 'base.spacing.x6',
    scale: '1',
  },
};

export const plainDotBadgeSx = {
  height: 'base.spacing.x2',
  width: 'base.spacing.x2',
  p: '0px',
};

export const getBadgeVariantSx = (variant: BadgeVariant) => {
  return {
    bg: `base.color.status.${variant}.bright`,

    '&:before': {
      bg: `base.color.status.${variant}.bright`,
    },
  };
};

export const textSx = {
  c: 'base.color.fixed.black.1000',
  position: 'relative',
};
