// @biom3/react/src/components/SpineMenu/styles.ts
export const spineMenuSx = {
  d: 'flex',
  flexDirection: 'column',
  w: '112px',
  h: '100%',
  flexGrow: 1,
  px: 'base.spacing.x2',
  py: 'base.spacing.x6',
  gap: 'base.spacing.x2',
  bg: 'base.color.neutral.700',
  boxShadow: 'base.shadow.200',
  overflowX: 'hidden',
  overflowY: 'auto',
};

export const spineMenuItemSx = {
  d: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  position: 'relative',
  gap: 'base.spacing.x2',
  py: 'base.spacing.x3',
  color: 'base.color.text.secondary',
  fill: 'base.color.text.secondary',
  borderRadius: 'base.borderRadius.x4',

  '&::before': {
    content: '""',
    position: 'absolute',
    left: '0px',
    top: '0px',
    w: '100%',
    h: '100%',
    borderRadius: 'base.borderRadius.x4',
    border: '1px solid transparent',
  },

  '&:hover, &.selected': {
    c: 'base.color.text.primary',
    fill: 'base.color.text.primary',
    bg: 'base.color.translucent.emphasis.200',
    boxShadow: 'base.shadow.300',
  },

  '&:hover::before, &.selected::before': {
    borderColor: 'base.color.translucent.standard.1000',
  },

  '&:active': {
    boxShadow: 'base.shadow.200',
  },

  '&:active::before': {
    borderWidth: '2px',
    borderColor: 'base.color.translucent.standard.1000',
  },
};

export const badgeSx = {
  position: 'absolute',
  top: '4px',
  left: '50%',
  transform: 'translateX(-50%)',
  ml: '10px',
};

export const iconSx = { width: 'base.icon.size.250', fill: 'inherit' };
