// @biom3/design-tokens/src/tokens/base.ts
import {
  Animation,
  BaseTokens,
  BorderRadius,
  Borders,
  Breakpoints,
  Font,
  Icons,
  Spacing,
  Text,
  ZLevels,
} from '../types';
import { onDarkColors, onDarkGradients, onDarkShadows } from './theme/onDark';
import {
  onLightColors,
  onLightGradients,
  onLightShadows,
} from './theme/onLight';

export const fontFaceStylesImport = `@import url("https://biome.immutable.com/hosted-assets/css/im-fonts-v4.css");`;

const border: Borders = {
  size: {
    600: '6px',
    500: '5px',
    400: '4px',
    300: '3px',
    200: '2px',
    100: '1px',
  },
} as const;

const icon: Icons = {
  size: {
    600: '64px',
    500: '48px',
    400: '32px',
    300: '24px',
    250: '20px',
    200: '16px',
    100: '12px',
  },
} as const;

const font: Font = {
  family: {
    heading: {
      primary: `'Suisse-Intl', sans-serif`,
      secondary: `'Suisse-Intl', sans-serif`,
    },
    body: {
      primary: `'Roboto', sans-serif`,
      secondary: `'Roboto Mono', monospace`,
    },
  },
  decoration: {
    underline: 'underline',
    lineThrough: 'line-through',
    none: 'none',
  },
  casing: {
    uppercase: 'uppercase',
    none: 'none',
  },
};

const text: Text = {
  body: {
    large: {
      regular: {
        fontSize: '20px',
        lineHeight: '28px',
        fontWeight: 400,
      },
      bold: {
        fontWeight: 500,
      },
    },
    medium: {
      regular: {
        fontSize: '16px',
        lineHeight: '24px',
        fontWeight: 400,
      },
      bold: {
        fontWeight: 500,
      },
    },
    small: {
      regular: {
        fontSize: '14px',
        lineHeight: '20px',
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    xSmall: {
      regular: {
        fontSize: '12px',
        lineHeight: '16px',
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    xxSmall: {
      regular: {
        fontSize: '10px',
        lineHeight: '14px',
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
  },
  caption: {
    medium: {
      regular: {
        fontSize: '14px',
        lineHeight: '20px',
        fontWeight: 400,
        casing: 'uppercase',
        letterSpacing: '2px',
      },
      bold: {
        fontSize: '14px',
        lineHeight: '20px',
        fontWeight: 700,
        casing: 'uppercase',
        letterSpacing: '2px',
      },
    },
    small: {
      regular: {
        fontSize: '10px',
        lineHeight: '14px',
        fontWeight: 400,
        casing: 'uppercase',
        letterSpacing: '2px',
      },
      bold: {
        fontSize: '10px',
        lineHeight: '14px',
        fontWeight: 700,
        casing: 'uppercase',
        letterSpacing: '2px',
      },
    },
  },
  heading: {
    xxLarge: {
      regular: {
        fontSize: '96px',
        lineHeight: '112px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    xLarge: {
      regular: {
        fontSize: '64px',
        lineHeight: '72px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    large: {
      regular: {
        fontSize: '40px',
        lineHeight: '48px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    medium: {
      regular: {
        fontSize: '28px',
        lineHeight: '32px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    small: {
      regular: {
        fontSize: '20px',
        lineHeight: '28px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
    xSmall: {
      regular: {
        fontSize: '16px',
        lineHeight: '24px',
        fontWeight: 600,
      },
      light: {
        fontWeight: 400,
      },
      bold: {
        fontWeight: 700,
      },
    },
  },
} as const;

const spacing: Spacing = {
  x1: '4px',
  x2: '8px',
  x3: '12px',
  x4: '16px',
  x5: '20px',
  x6: '24px',
  x7: '28px',
  x8: '32px',
  x9: '36px',
  x10: '40px',
  x11: '44px',
  x12: '48px',
  x13: '52px',
  x14: '56px',
  x15: '60px',
  x16: '64px',
  x17: '68px',
  x18: '72px',
  x19: '76px',
  x20: '80px',
  x21: '84px',
  x22: '88px',
  x23: '92px',
  x24: '96px',
  x25: '100px',
  x26: '104px',
  x27: '108px',
  x28: '112px',
  x29: '116px',
  x30: '120px',
  x31: '124px',
  x32: '128px',
  x33: '132px',
  x34: '136px',
  x35: '140px',
  x36: '144px',
  x37: '148px',
  x38: '152px',
  x39: '156px',
  x40: '160px',
  x41: '164px',
  x42: '168px',
  x43: '172px',
  x44: '176px',
  x45: '180px',
  x46: '184px',
  x47: '188px',
  x48: '192px',
  x49: '196px',
  x50: '200px',
} as const;

const borderRadius: BorderRadius = {
  x1: '2px',
  x2: '4px',
  x3: '6px',
  x4: '8px',
  x5: '10px',
  x6: '12px',
  x7: '14px',
  x8: '16px',
  x9: '18px',
  x10: '20px',
  x11: '22px',
  x12: '24px',
  x13: '26px',
  x14: '28px',
  x15: '30px',
  x16: '32px',
  x17: '34px',
  x18: '36px',
  x19: '38px',
  x20: '40px',
  x21: '42px',
  x22: '44px',
  x23: '46px',
  x24: '48px',
  x25: '50px',
} as const;

const breakpoint: Breakpoints = {
  small: 450,
  medium: 768,
  large: 1024,
  xLarge: 1280,
  xxLarge: 1536,
  xxxLarge: 1920,
} as const;

const zLevel: ZLevels = {
  tooltip: 12,
  toast: 11,
  popover: 10,
  modal: 9,
  drawer: 8,
} as const;

const motion: Animation = {
  normal: {
    gentle: {
      cssDuration: '0.4s',
      jsDuration: 0.4,
      cssEase: 'ease-in-out',
      jsEase: 'easeInOut',
    },
    fast: {
      cssDuration: '0.25s',
      jsDuration: 0.25,
      cssEase: 'ease-in-out',
      jsEase: 'easeInOut',
    },
  },
  bounce: {
    gentle: {
      cssDuration: '0.6s',
      jsDuration: 0.6,
      cssEase: 'cubic-bezier(0.47, 1.74, 0.41, 0.8)',
      jsEase: [0.47, 1.74, 0.41, 0.8],
    },
    fast: {
      cssDuration: '0.45s',
      jsDuration: 0.45,
      cssEase: 'cubic-bezier(0.47, 1.74, 0.41, 0.8)',
      jsEase: [0.47, 1.74, 0.41, 0.8],
    },
  },
};

export const base: BaseTokens = {
  colorMode: 'darkOnLight',
  color: onLightColors,
  gradient: onLightGradients,
  shadow: onLightShadows,
  spacing,
  borderRadius,
  font,
  text,
  breakpoint,
  icon,
  border,
  zLevel,
  motion,
};

export const onDarkBase: BaseTokens = {
  colorMode: 'lightOnDark',
  color: onDarkColors,
  gradient: onDarkGradients,
  shadow: onDarkShadows,
  spacing,
  borderRadius,
  font,
  text,
  breakpoint,
  icon,
  border,
  zLevel,
  motion,
};
