// @biom3/react/src/components/Checkbox/styles.ts
import { BiomeTheme } from '../../types/theme';

export const containerStyles = {
  width: 'base.spacing.x5',
  height: 'base.spacing.x5',
  position: 'relative',
};

export const inputCssStyles = (theme: BiomeTheme) => ({
  '&:disabled + .checkContainer': {
    boxShadow: 'none',
  },

  '&:checked + .checkContainer': {
    borderWidth: theme.base.border.size[200],
    borderColor: theme.base.color.text.primary,
  },

  '&:hover:not(:disabled) + .checkContainer': {
    borderWidth: theme.base.border.size[200],
  },

  '&:checked:hover:not(:disabled) + .checkContainer': {
    borderColor: theme.base.color.text.secondary,
    borderWidth: theme.base.border.size[100],
  },

  '&:checked + .checkContainer > .checkIcon': {
    opacity: '1',
  },
});

export const checkContainerSx = {
  borderStyle: 'solid',
  borderWidth: 'base.border.size[100]',
  borderColor: 'base.color.text.secondary',
  borderRadius: 'base.borderRadius.x4',
  boxShadow: 'base.shadow[100]',
  width: 'base.spacing.x5',
  height: 'base.spacing.x5',
  d: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
};

export const tickSx = {
  fill: 'base.color.text.primary',
  pointerEvents: 'none',
  opacity: '0',
};
