// @biom3/react/src/components/ProgressBar/shared.ts
import { ReactNode } from 'react';

import { AnimationKind } from '../../types/motion';
import { MakeResponsive, StandardComponentWithProps } from '../../types/shared';

export const PROGRESS_BAR_VARIANTS = ['determinate', 'indeterminate'] as const;
export const PROGRESS_BAR_SIZES = ['small', 'medium'] as const;
export const DEFAULT_PROGRESS_BAR_SIZE = PROGRESS_BAR_SIZES[1];
export const PROGRESS_BAR_COLOR_VARIANTS = [
  'guidance',
  'brand',
  'inverted',
  'gradient',
] as const;
export const DEFAULT_PROGRESS_BAR_COLOR_VARIANT =
  PROGRESS_BAR_COLOR_VARIANTS[0];

export type ProgressBarVariant = (typeof PROGRESS_BAR_VARIANTS)[number];
export type ProgressBarSize = (typeof PROGRESS_BAR_SIZES)[number];
export type ProgressBarColorVariant =
  (typeof PROGRESS_BAR_COLOR_VARIANTS)[number];

export type ProgressBarProps = StandardComponentWithProps<
  HTMLDivElement,
  {
    children?: ReactNode;
    colorVariant?: ProgressBarColorVariant;
    size?: MakeResponsive<ProgressBarSize>;
    value?: number;
    animationKind?: AnimationKind;
  } & (
    | { variant: 'determinate'; value: number }
    | { variant: 'indeterminate'; value?: never }
  )
>;
