// @biom3/react/src/components/Logo/Logo.tsx
import { ReactElement } from 'react';

import { SingleVariantIconProps } from '../../types/icon';
import { DomPropsWithDomRef } from '../../types/shared';
import { DocsHorizontalLockup } from './DocsHorizontalLockup';
import { DocsHorizontalLockupCompact } from './DocsHorizontalLockupCompact';
import { DocsVerticalLockup } from './DocsVerticalLockup';
import { GameStopSymbol } from './GameStopSymbol';
import { HubHorizontalLockup } from './HubHorizontalLockup';
import { HubHorizontalLockupCompact } from './HubHorizontalLockupCompact';
import { HubVerticalLockup } from './HubVerticalLockup';
import { ImmutableHorizontalLockup } from './ImmutableHorizontalLockup';
import { ImmutableSymbol } from './ImmutableSymbol';
import { ImmutableVerticalLockup } from './ImmutableVerticalLockup';
import { MagicSymbol } from './MagicSymbol';
import { MetaMaskSymbol } from './MetaMaskSymbol';
import { PassportHorizontalLockup } from './PassportHorizontalLockup';
import { PassportHorizontalLockupCompact } from './PassportHorizontalLockupCompact';
import { PassportSymbol } from './PassportSymbol';
import { PassportSymbolOutlined } from './PassportSymbolOutlined';
import { PassportVerticalLockup } from './PassportVerticalLockup';
import { WalletConnectSymbol } from './WalletConnectSymbol';

export const lockupLogos = {
  ImmutableHorizontalLockup,
  ImmutableVerticalLockup,
  PassportHorizontalLockup,
  PassportHorizontalLockupCompact,
  PassportVerticalLockup,
  HubHorizontalLockup,
  HubHorizontalLockupCompact,
  HubVerticalLockup,
  DocsHorizontalLockup,
  DocsHorizontalLockupCompact,
  DocsVerticalLockup,
} as const;

export const immutableSymbolLogos = {
  ImmutableSymbol,
  PassportSymbol,
  PassportSymbolOutlined,
} as const;

export const otherSymbolLogos = {
  MetaMaskSymbol,
  WalletConnectSymbol,
  MagicSymbol,
  GameStopSymbol,
} as const;

export const allLogos = {
  ...lockupLogos,
  ...immutableSymbolLogos,
  ...otherSymbolLogos,
};

export type AllLogoKeys = keyof typeof allLogos;

export type LogoProps = SingleVariantIconProps & {
  logo: AllLogoKeys;
};

export function Logo<RC extends ReactElement | undefined = undefined>({
  logo,
  ...props
}: RC extends undefined
  ? DomPropsWithDomRef<'svg'> & LogoProps
  : LogoProps & { rc: RC }) {
  const LogoComponent = allLogos[logo];
  return <LogoComponent {...props} />;
}

Logo.displayName = 'Logo';
