// @biom3/design-tokens/src/tokens/theme/onDark.ts
import { Colors, Gradients, Shadows } from '../../types';

export const onDarkColors: Colors = {
  brand: {
    1: '#F3F3F3',
    2: '#131313',
    3: '#E0E0E0',
    4: '#B6B6B6',
    5: '#36E0F8',
    6: '#1F8290',
  },
  neutral: {
    500: '#575757',
    600: '#444444',
    700: '#323232',
    800: '#1F1F1F',
    900: '#131313',
    1000: '#0D0D0D',
  },
  accent: {
    1: '#F191FA',
    2: '#36E0F8',
    3: '#6FD6F2',
    4: '#61ADEB',
    5: '#99A8EA',
    6: '#E4F4F7',
    7: '#F8D1DE',
    8: '#F2B7DA',
  },
  translucent: {
    emphasis: {
      500: '#F3F3F333',
      400: '#F3F3F329',
      300: '#F3F3F31F',
      200: '#F3F3F314',
      100: '#F3F3F30A',
    },
    standard: {
      1000: '#F3F3F3FF', // rgba(243, 243, 243, 1)
      900: '#F3F3F3E0', // rgba(243, 243, 243, 0.88)
      800: '#F3F3F3CC', // rgba(243, 243, 243, 0.80)
      700: '#F3F3F3A3', // rgba(243, 243, 243, 0.64)
      600: '#F3F3F37A', // rgba(243, 243, 243, 0.48)
      500: '#F3F3F33D', // rgba(243, 243, 243, 0.24)
      400: '#F3F3F329', // rgba(243, 243, 243, 0.16)
      300: '#F3F3F31F', // rgba(243, 243, 243, 0.12)
      200: '#F3F3F314', // rgba(243, 243, 243, 0.08)
      150: '#F3F3F30F', // rgba(243, 243, 243, 0.06)
      100: '#F3F3F30A', // rgba(243, 243, 243, 0.04)
      50: '#F3F3F305', // rgba(243, 243, 243, 0.02)
    },
    inverse: {
      1000: '#0D0D0DFF',
      900: '#0D0D0DE6',
      800: '#0D0D0DCC',
      700: '#0D0D0DA3',
      600: '#0D0D0D7A',
      500: '#0D0D0D3D',
      400: '#0D0D0D29',
      300: '#0D0D0D1F',
      200: '#0D0D0D14',
      100: '#0D0D0D0A',
    },
  },
  status: {
    guidance: {
      bright: '#36E0F8',
      dim: '#94E9F5 ',
    },
    success: {
      bright: '#ABF790',
      dim: '#D5FAC7',
    },
    attention: {
      bright: '#FAFD7E',
      dim: '#F6F8B9',
    },
    fatal: {
      bright: '#FF637F',
      dim: '#F9ABB9',
    },
    destructive: {
      bright: '#E01A3D',
      dim: '#FF637F',
    },
  },
  text: {
    primary: 'base.color.brand.1',
    secondary: 'base.color.brand.4',
    link: {
      primary: 'base.color.accent.1',
      secondary: 'base.color.brand.1',
    },
    inverse: {
      primary: 'base.color.brand.2',
      secondary: '#464646',
    },
  },
  fixed: {
    black: {
      1000: 'base.brand.2',
      100: '#13131314',
    },
    white: {
      1000: 'base.brand.1',
      100: '#F3F3F314',
    },
  },
} as const;

export const onDarkShadows: Shadows = {
  100: '0px 1px 2px rgba(0, 0, 0, 0.16), 0px 1px 4px 1px rgba(0, 0, 0, 0.32)',
  200: '0px 2px 6px 2px rgba(0, 0, 0, 0.16), 0px 1px 2px rgba(0, 0, 0, 0.32)',
  300: '0px 4px 8px 4px rgba(0, 0, 0, 0.16), 0px 1px 4px rgba(0, 0, 0, 0.32)',
  400: '0px 6px 10px 4px rgba(0, 0, 0, 0.16), 0px 2px 4px rgba(0, 0, 0, 0.32)',
  500: '0px 8px 12px 6px rgba(0, 0, 0, 0.16), 0px 4px 4px rgba(0, 0, 0, 0.32)',
} as const;

export const onDarkGradients: Gradients = {
  1: {
    spectrum: 'linear-gradient(105.76deg, #F191FA -46.96%, #83E3F0 99.99%)',
    blendMode: 'normal',
    svgDefs: `
    <linearGradient id="gradient-1-singleLayer" x1="0" y1="0" x2="24" y2="0.214702"
      gradientUnits="userSpaceOnUse">
      <stop stop-color="#C9A2F6" />
      <stop offset="1" stop-color="#83E3F0" />
    </linearGradient>`,
    svgDefIds: 'gradient-1-singleLayer',
  },
  2: {
    spectrum: `radial-gradient(96.87% 65.57% at 54.34% 93.79%, #FCF5EE 0%, rgba(236, 190, 225, 0) 71.51%), radial-gradient(73.48% 242.97% at 26.54% 26.42%, #A3EEF8 0%, #A4DCF5 17.71%, #A6AEEC 38.02%, #ECBEE1 100%)`,
    blendMode: 'normal, normal',
    svgDefs: `
    <radialGradient id="gradient-2-bottomLayer" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse"
      gradientTransform="translate(4.7777 4.75595) rotate(44.9817) scale(18.6995 39.6574)">
      <stop stop-color="#A3EEF8" />
      <stop offset="0.177083" stop-color="#A4DCF5" />
      <stop offset="0.380208" stop-color="#A6AEEC" />
      <stop offset="1" stop-color="#ECBEE1" />
    </radialGradient>
    <radialGradient id="gradient-2-topLayer" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse"
      gradientTransform="translate(9.7806 16.8827) rotate(84.265) scale(9.68552 18.5286)">
      <stop stop-color="#FCF5EE" />
      <stop offset="0.715135" stop-color="#ECBEE1" stop-opacity="0" />
    </radialGradient>`,
    svgDefIds: `gradient-2-bottomLayer, gradient-2-topLayer`,
  },
  3: {
    spectrum: `radial-gradient(100% 100% at 24.5% 100%, rgba(106, 160, 242, 0.9) 0%, rgba(198, 229, 247, 0) 100%), linear-gradient(81.66deg, rgba(124, 246, 251, 0.48) 9.38%, rgba(254, 253, 248, 0.7) 98.47%), radial-gradient(4009.6% 10453.55% at 0% 0%, #7EF8FE 0%, #7ABFEC 30.21%, #AEC3ED 45.54%, #DEF3F6 61.46%, #FAF9F3 100%)`,
    blendMode: 'normal, normal, normal',
    svgDefs: `
    <radialGradient id="gradient-3-bottomLayer" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse"
      gradientTransform="rotate(60.6728) scale(36.75 1550.85)">
      <stop stop-color="#7EF8FE" />
      <stop offset="0.302083" stop-color="#7ABFEC" />
      <stop offset="0.455396" stop-color="#AEC3ED" />
      <stop offset="0.614583" stop-color="#DEF3F6" />
      <stop offset="1" stop-color="#FAF9F3" />
    </radialGradient>
    <linearGradient id="gradient-3-middleLayer" x1="-1.26123e-07" y1="4.8" x2="24" y2="2.16"
      gradientUnits="userSpaceOnUse">
      <stop stop-color="#7CF6FB" stop-opacity="0.48" />
      <stop offset="1" stop-color="#FEFDF8" stop-opacity="0.7" />
    </linearGradient>
    <radialGradient id="gradient-3-topLayer" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse"
      gradientTransform="translate(4.41 18) rotate(-52.9472) scale(22.5541)">
      <stop stop-color="#6AA0F2" />
      <stop offset="1" stop-color="#C6E5F7" stop-opacity="0" />
    </radialGradient>`,
    svgDefIds: `gradient-3-bottomLayer,
      gradient-3-middleLayer,
      gradient-3-topLayer`,
  },
} as const;
