// @imtbl/contracts/dist/typechain-types/factories/contracts/token/erc721/x/IMintable__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMintable__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "to",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "quantity",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "mintingBlob",
                type: "bytes",
            },
        ],
        name: "mintFor",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IMintable__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IMintable__factory = IMintable__factory;
IMintable__factory.abi = _abi;
