// @biom3/react/src/types/sxThemableProperties.ts
import {
  BorderRadiusTokenPaths,
  BorderTokenPaths,
  ColorTokenPaths,
  FontTokenPaths,
  GradientTokenPaths,
  IconTokenPaths,
  LooseAutocomplete,
  MotionTokenPaths,
  ShadowTokenPaths,
  SpacingTokenPaths,
  TextTokenPaths,
} from '@biom3/design-tokens';

import { MakeValidSxValue, ValidSxValues } from './sxProps';

// These are the properties that can be themed in the sx prop. They correspond to
// specific paths of design tokens in the design-tokens package.
export type SpacingProperties =
  | 'm'
  | 'margin'
  | 'marginTop'
  | 'mt'
  | 'marginRight'
  | 'mr'
  | 'marginBottom'
  | 'mb'
  | 'marginLeft'
  | 'ml'
  | 'marginX'
  | 'mx'
  | 'marginY'
  | 'my'
  | 'p'
  | 'padding'
  | 'paddingTop'
  | 'pt'
  | 'paddingRight'
  | 'pr'
  | 'paddingBottom'
  | 'pb'
  | 'paddingLeft'
  | 'pl'
  | 'paddingX'
  | 'px'
  | 'paddingY'
  | 'py'
  | 'scrollMargin'
  | 'scrollMarginTop'
  | 'scrollMarginRight'
  | 'scrollMarginBottom'
  | 'scrollMarginLeft'
  | 'scrollMarginX'
  | 'scrollMarginY'
  | 'scrollPadding'
  | 'scrollPaddingTop'
  | 'scrollPaddingRight'
  | 'scrollPaddingBottom'
  | 'scrollPaddingLeft'
  | 'scrollPaddingX'
  | 'scrollPaddingY'
  | 'top'
  | 'right'
  | 'bottom'
  | 'left'
  | 'gridGap'
  | 'gridColumnGap'
  | 'gridRowGap'
  | 'gap'
  | 'columnGap'
  | 'rowGap'
  | 'width'
  | 'w'
  | 'minw'
  | 'maxw'
  | 'minWidth'
  | 'maxWidth'
  | 'columnWidth'
  | 'height'
  | 'h'
  | 'minh'
  | 'maxh'
  | 'minHeight'
  | 'maxHeight'
  | 'flexBasis'
  | 'size'
  | 'letterSpacing';

export type ColorProperties =
  | 'color'
  | 'c'
  | 'bg'
  | 'background'
  | 'backgroundColor'
  | 'borderColor'
  | 'caretColor'
  | 'outlineColor'
  | 'textDecorationColor'
  | 'fill'
  | 'stroke'
  | 'borderTopColor'
  | 'borderRightColor'
  | 'borderBottomColor'
  | 'borderLeftColor'
  | 'columnRuleColor'
  | 'bgc';

export type GradientProperties =
  | 'bg'
  | 'background'
  | 'color'
  | 'c'
  | 'backgroundImage'
  | 'bgi';

export type ShadowProperties = 'boxShadow' | 'textShadow';

export type RadiusProperties =
  | 'borderRadius'
  | 'borderTopLeftRadius'
  | 'borderTopRightRadius'
  | 'borderBottomLeftRadius'
  | 'borderBottomRightRadius'
  | 'brad'
  | 'bradtl'
  | 'bradtr'
  | 'bradbl'
  | 'bradbr';

export type FontProperties = 'fontFamily' | 'textDecoration' | 'textTransform';

export type TextProperties = 'fontSize' | 'fontWeight' | 'lineHeight';

export type IconSizeProperties = SpacingProperties;

export type BorderProperties =
  | 'borderTop'
  | 'borderRight'
  | 'borderBottom'
  | 'borderLeft'
  | 'borderWidth'
  | 'borderTopWidth'
  | 'borderRightWidth'
  | 'borderBottomWidth'
  | 'borderLeftWidth';

export type MotionProperties =
  | 'transitionDuration'
  | 'transitionTimingFunction'
  | 'animationDuration'
  | 'animationTimingFunction';

export type GetNarrowedDesignTokens<T> = MakeValidSxValue<
  LooseAutocomplete<
    NonNullable<
      | (T extends SpacingProperties ? SpacingTokenPaths : undefined)
      | (T extends ColorProperties ? ColorTokenPaths : undefined)
      | (T extends GradientProperties ? GradientTokenPaths : undefined)
      | (T extends ShadowProperties ? ShadowTokenPaths : undefined)
      | (T extends RadiusProperties ? BorderRadiusTokenPaths : undefined)
      | (T extends FontProperties ? FontTokenPaths : undefined)
      | (T extends TextProperties ? TextTokenPaths : undefined)
      | (T extends IconSizeProperties ? IconTokenPaths : undefined)
      | (T extends BorderProperties ? BorderTokenPaths : undefined)
      | (T extends MotionProperties ? MotionTokenPaths : undefined)
      | (T extends
          | SpacingProperties
          | ColorProperties
          | GradientProperties
          | ShadowProperties
          | RadiusProperties
          | FontProperties
          | TextProperties
          | IconSizeProperties
          | BorderProperties
          | MotionProperties
          ? undefined
          : ValidSxValues)
    >
  >
>;
